### PR TITLE
[IMP] core: void model translation

### PR DESCRIPTION
**Pull request by HydrionBurst**
## before 1st commit:
there is no way to void model translations. The api `write` and `update_field_translations` can only overwrite the translations for the specified languages.

## after 1st commit:
non-en_US empty str translation for `update_field_translations` means void the translation for the specified language and fallback to en_US value

## before 2nd commit:
model translation:
when the user changes the translation for the current language in the
translation dialog, the value in the form view is also updated and marked
modified.
But if the user wants to invalidate the translation and set the translation to
`false`, the form view value is also updated to `false`, if the user saves by
accident, all translations are lost.

model_terms translation:
when translations are modified in the translation dialog, the field value will
not be changed

## after 2nd commit:
form view data will be reloaded after translations changed in the translation
dialog

Description of the issue/feature this PR addresses:

Current behavior before PR:

Desired behavior after PR is merged:

---
I confirm I have signed the CLA and read the PR guidelines at www.odoo.com/submit-pr

### CODE REVIEW

File: odoo/addons/base/models/ir_cron.py [hunk 1]
```diff
@@ -50,7 +50,7 @@ class ir_cron(models.Model):
     ir_actions_server_id = fields.Many2one(
         'ir.actions.server', 'Server action',
         delegate=True, ondelete='restrict', required=True)
-    cron_name = fields.Char('Name', related='ir_actions_server_id.name', store=True, readonly=False)
+    cron_name = fields.Char('Name', related='ir_actions_server_id.name', readonly=False)
```
Comment (rco-odoo) [inline odoo/addons/base/models/ir_cron.py:53]: You cannot do that on a stable version.
Moreover, I don't see how it relates to the issue you are fixing.

File: odoo/addons/base/tests/test_translate.py [hunk 2]
```diff
@@ -687,23 +687,31 @@ def _test_create_empty(self, empty_value):
         self.assertEqual(group.with_context(lang='fr_FR').comment, 'French comment 2')
         self.assertEqual(group.with_context(lang='en_US').comment, 'French comment')
 
-    def test_field_selection(self):
-        """ Test translations of field selections. """
+    def test_update_field_translations(self):
```
Comment (rco-odoo) [inline odoo/addons/base/tests/test_translate.py:690]: Why did you remove the method `test_field_selection` ?
Please put it back (unless there is a good reason.)

File: odoo/addons/base/tests/test_translate.py [hunk 3]
```diff
@@ -687,23 +687,31 @@ def _test_create_empty(self, empty_value):
...
-
-        fg = self.env['ir.model'].fields_get(['state'])
-        self.assertEqual(fg['state']['selection'], field.selection)
 
-        fg = self.env['ir.model'].with_context(lang='fr_FR').fields_get(['state'])
-        self.assertEqual(fg['state']['selection'],
-                         [('manual', 'Custo'), ('base', 'Pas touche!')])
+        langs = self.env['res.lang'].get_installed()
+        self.assertEqual([('en_US', 'English (US)'), ('fr_FR', 'French / Français')], langs,
+                         "Test did not started with expected languages")
```
Comment (rco-odoo) [inline odoo/addons/base/tests/test_translate.py:695]: Most tests in this class rely on this, no need to check it.  Please remove those lines.

File: odoo/addons/base/tests/test_translate.py [hunk 4]
```diff
@@ -687,23 +687,31 @@ def _test_create_empty(self, empty_value):
...
+        categoryEN = self.category.with_context(lang='en_US')
+        categoryFR = self.category.with_context(lang='fr_FR')
+
+        self.category.update_field_translations('name', {'en_US': 'English Name', 'fr_FR': 'French Name'})
+        self.assertEqual(categoryFR.name, 'French Name')
+        self.assertEqual(categoryEN.name, 'English Name')
```
Comment (rco-odoo) [inline odoo/addons/base/tests/test_translate.py:701]: Only for the sake of readability, I would swap those two lines to make the order consistent everywhere (we always check English then French.)

File: odoo/addons/base/tests/test_translate.py [hunk 5]
```diff
@@ -687,23 +687,31 @@ def _test_create_empty(self, empty_value):
...
+        self.assertEqual(categoryFR.name, 'English Name')
+        categoryEN.name = 'English Name 2'
+        self.assertEqual(categoryFR.name, 'English Name 2')
+
+        # cannot void en_US and empty string is considered as a valid value for translation
+        self.category.update_field_translations('name', {'en_US': 'English Name', 'fr_FR': 'French Name'})
+        self.category.update_field_translations('name', {'en_US': ''})
+        self.assertEqual(categoryEN.name, '')
+        self.assertEqual(categoryFR.name, 'French Name')
```
Comment (rco-odoo) [inline odoo/addons/base/tests/test_translate.py:714]: I am not 100% sure about this one.
@odony what do you think?

File: odoo/models.py [hunk 6]
```diff
@@ -3025,25 +3025,31 @@ def _update_field_translations(self, field_name, translations, digest=None):
...
+            return self.mapped(related_path)._update_field_translations(field_name, translations, digest)
+        self.check_access_rights('write')
+        self.check_field_access_rights('write', [field_name])
+        self.check_access_rule('write')
+
         if field.translate is True:
-            for lang, translation in translations.items():
-                if translation is not None:
-                    self.with_context(lang=lang)[field_name] = translation
+            # empty string translation and Falsy translation are used to void the corresponding translation
```
Comment (rco-odoo) [inline odoo/models.py:3042]: Nitpicking: empty strings are falsy, hence
```suggestion
            # falsy translation are used to void the corresponding translation
```

File: odoo/models.py [hunk 7]
```diff
@@ -3025,25 +3025,31 @@ def _update_field_translations(self, field_name, translations, digest=None):
...
+            translations = {
+                lang: None if not translation and lang != 'en_US' else str(translation)
```
Comment (rco-odoo) [inline odoo/models.py:3044]: If `translation` is `False`, `str(translation)` is `"False"` :smile: 
```suggestion
                lang: None if (not translation and lang != 'en_US') else (translation or "")
```
Make the test cover that case for English and non-English, please.

File: odoo/models.py [hunk 8]
```diff
@@ -3025,25 +3025,31 @@ def _update_field_translations(self, field_name, translations, digest=None):
...
+                for lang, translation in translations.items()
+            }
+            self.invalidate_recordset([field_name])
+            self._cr.execute(f'''
+                UPDATE {self._table} SET {field_name} = jsonb_strip_nulls({field_name} || %s)
```
Comment (rco-odoo) [inline odoo/models.py:3049]: Ahem, this updates **all** records in the table!  I think you forgot the WHERE clause :wink:

**Review by rco-odoo — CHANGES_REQUESTED** [review]

File: odoo/models.py [hunk 9]
```diff
@@ -3025,25 +3025,32 @@ def _update_field_translations(self, field_name, translations, digest=None):
...
+
         if field.translate is True:
-            for lang, translation in translations.items():
-                if translation is not None:
-                    self.with_context(lang=lang)[field_name] = translation
+            # falsy translation are used to void the corresponding translation
+            translations = {
+                lang: None if (not translation and lang != 'en_US') else translation
+                for lang, translation in translations.items()
+                if isinstance(translation, str)
```
Comment (rco-odoo) [inline odoo/models.py:3046]: So `translation=False` is simply be ignored?  That feels awkward.

File: odoo/addons/base/tests/test_translate.py [hunk 10]
```diff
@@ -687,6 +687,28 @@ def _test_create_empty(self, empty_value):
...
+        self.category.update_field_translations('name', {'en_US': 'English Name', 'fr_FR': 'French Name'})
+        self.assertEqual(categoryEN.name, 'English Name')
+        self.assertEqual(categoryFR.name, 'French Name')
+
+        # void fr_FR translation and fallback to en_US
+        self.category.update_field_translations('name', {'fr_FR': ''})
+        self.assertEqual(categoryEN.name, 'English Name')
+        self.assertEqual(categoryFR.name, 'English Name')
+        categoryEN.name = 'English Name 2'
+        self.assertEqual(categoryFR.name, 'English Name 2')
```
Comment (rco-odoo) [inline odoo/addons/base/tests/test_translate.py:704]: Can you please add similar test cases with `False` instead of `''` ?

File: odoo/models.py [hunk 11]
```diff
@@ -3025,25 +3025,32 @@ def _update_field_translations(self, field_name, translations, digest=None):
...
+            return self.mapped(related_path)._update_field_translations(field_name, translations, digest)
+        self.check_access_rights('write')
+        self.check_field_access_rights('write', [field_name])
+        self.check_access_rule('write')
+
         if field.translate is True:
-            for lang, translation in translations.items():
-                if translation is not None:
-                    self.with_context(lang=lang)[field_name] = translation
+            # falsy translation are used to void the corresponding translation
```
Comment (rco-odoo) [inline odoo/models.py:3042]:
```suggestion
            # falsy values are used to void the corresponding translation
```

File: odoo/models.py [hunk 12]
```diff
@@ -3025,25 +3025,32 @@ def _update_field_translations(self, field_name, translations, digest=None):
...
+            translations = {
+                lang: None if (not translation and lang != 'en_US') else (translation or '')
+                for lang, translation in translations.items()
+                if isinstance(translation, str) or not translation
```
Comment (rco-odoo) [inline odoo/models.py:3046]: Given that we expect `translation` to be a string or a falsy value, this condition is useless.  It is actually worse, since it discards values that should be considered as errors.

File: odoo/models.py [hunk 13]
```diff
@@ -3025,25 +3025,32 @@ def _update_field_translations(self, field_name, translations, digest=None):
...
+        self.check_field_access_rights('write', [field_name])
+        self.check_access_rule('write')
+
         if field.translate is True:
-            for lang, translation in translations.items():
-                if translation is not None:
-                    self.with_context(lang=lang)[field_name] = translation
+            # falsy translation are used to void the corresponding translation
+            translations = {
+                lang: None if (not translation and lang != 'en_US') else (translation or '')
```
Comment (rco-odoo) [inline odoo/models.py:3044]: This condition is a bit difficult to read.  What about this?
```suggestion
                lang: (translation or "") if lang == 'en_US' else (translation or None)
```

**Comment by HydrionBurst** [comment]
@rco-odoo
I rethink about the api on weekends.
I prefer we can keep the behavior for translating to `''` for the backend api, because the `''` is still considered as a valid translation in Cache/Database
On the other hand, I want to put the 'void translations' logic in the UI. The UI will send `{'fr_FR': false}` to the backend if the translation is changed to `""`

**Comment by rco-odoo** [comment]
> As a result `model.with_context(lang='fr_FR').name = False` means void the field value `model.with_context(lang='fr_FR').name = ''` means set the fr_FR translation to `''` `model.update_field_translations('name', {'fr_FR': False})` means void fr_FR translation `model.update_field_translations('name', {'fr_FR': ''})` means set the fr_FR translation to `''`

This looks fine to me.  I even suggest to add these 4 lines somewhere in the documentation of method `update_field_translations` (docstring or comments), because it really clarifies the APIs in those corner cases.  And they must be covered by tests, of course :wink:

**Review by rco-odoo — APPROVED** [review]
This looks good to me. @odony

**Comment by HydrionBurst** [comment]
@odony 
Thanks for the feedback.
I hold an different idea.
1. To keep everything simple, as the name of the API `update_field_translations`, I hope it can focus on translations.
Within the scope of this function, the content of the field value should not be changed.
nullfying the field value/checking the required attribute mean the field value is changed, and we already have `write` api to do the job.
And the user can send False to `write` from the list/form view
2. It would be hard to define the result of `model.update_field_translations('name', {'en_US': False, 'fr_FR': 'new French value'}`
it may equal to
	a) `model_EN.name = False` then `model_FR.name = 'new French value'` => DB: `{'en_US': 'new French value', 'fr_FR': 'new French value'}`
	b) `model_FR.name = 'new French value'` then `model_EN.name = False` => DB: `null`
3. The current behavior is the same as 15.x, a `False` en_US translation means untranslated. When read, it will fallback to the source value. And source value is equal to the en_US value.

I suggest we keep it this way, if we get the feedback from our customers, we can improve it later.
cc @rco-odoo 

BTW there is a bad UX in both 15.x from time ago and this one
1. go fr_FR
2. open a translation dialog
3. remove fr_FR translation
5. save
6. in the form view, the field becomes empty
7. if you click `Save` in the form view, the UI will write `False` to the field -> lose all translations

I will take some time to solve this problem(maybe in another PR)

**Comment by mart-e** [comment]
> BTW there is a bad UX in both 15.x from time ago and this one

Also we have something strange with the required fields:

- db with en, fr and nl; user in fr on a translated record
- open translation wizard
- remove the french translation
  -> error field required
- refresh the page
 -> the required field has an empty value
in 15.0, we also have this issue but we fallback on the `src` value instead (which is odd as well)

**Comment by HydrionBurst** [comment]
> > BTW there is a bad UX in both 15.x from time ago and this one
> 
> Also we have something strange with the required fields:
> 
> * db with en, fr and nl; user in fr on a translated record
> * open translation wizard
> * remove the french translation
>   -> error field required
> * refresh the page
>   -> the required field has an empty value
>   in 15.0, we also have this issue but we fallback on the `src` value instead (which is odd as well)

It is because the current translation dialog may use empty string `''` for the new translation.
When it is saved, the database value is updated to `{'en_US': 'English', 'fr_FR': ''}`
After refresh the form view, the fr_FR value becomes `''` which is treated as a valid translation.

After this PR, the translation dialog will use `False` for the new translation,
When it is saved, the database value will be updated to `{'en_US': 'English value'}`
After refresh the form view, the fr_FR value is fallback to 'English value', like 15.x

**Comment by HydrionBurst** [comment]
Hi, @odoo/rd-framework-js could you review the js part of this PR for the translation dialog?

**Review by aab-odoo — COMMENTED** [review]
@HydrionBurst Isn't it a bit brutal to reload the webclient when a translated term is updated? As far as I understand, this is because you want to update the translated value, not the value itself. I think it wouldn't be difficult to update the js structure encoding loaded translations.

**Comment by aab-odoo** [comment]
I'm only talking about the second commit which concerns the edition from the dialog. I still think we get avoid the page reload. You can force a reload of the record (and keep changes if there are)

**Comment by aab-odoo** [comment]
> Is there any api can reload the form view data of the record without reloading the whole page?

From the form controller, you can access the record datapoint via `this.model.root`. From a field component, it's simply `this.props.record`. Then if you have the record, you can do `record.load({}, { keepChanges: true });`

File: addons/web/static/src/views/fields/translation_button.js [hunk 14]
```diff
@@ -50,7 +50,7 @@ export function useTranslationDialog() {
             resModel: resModel,
             userLanguageValue: record.data[fieldName] || "",
             isComingFromTranslationAlert: false,
-            updateField,
+            record: record,
```
Comment (aab-odoo) [inline addons/web/static/src/views/fields/translation_button.js:53]: suggestion: give a `onSave` callback in props instead of the record (to avoid leaking implem details from a component to another):
```js
async onSave() {
    await record.load({}, { keepChanges: true });
    record.model.notify();
}
``` 

and call `this.props.onSave();` where you currently reload the record in the dialog.

**Review by aab-odoo — APPROVED** [review]

**Comment by rco-odoo** [comment]
@robodoo r+

**Comment by rco-odoo** [comment]
@robodoo rebase-ff

**Comment by HydrionBurst** [comment]
@robodoo r+
